File: api/functionapp/user/Services/UserService.cs
using System.Security.Cryptography;
using System.Text;
using UserFunction.DTOs;
using UserFunction.Exceptions;
using UserFunction.Models;
using UserFunction.Repositories;
using Microsoft.Extensions.Logging;

namespace UserFunction.Services;

public class UserService : IUserService
{
    private readonly IUserRepository _userRepository;
    private readonly ILogger<UserService> _logger;

    public UserService(IUserRepository userRepository, ILogger<UserService> logger)
    {
        _userRepository = userRepository;
        _logger = logger;
    }

    public UserResponse CreateUser(CreateUserRequest request)
    {
        // Check if user already exists
        if (_userRepository.ExistsByEmail(request.Email))
        {
            throw new UserAlreadyExistsException($"User with email {request.Email} already exists");
        }

        // Create new user
        var user = new User
        {
            Id = Guid.NewGuid().ToString(),
            Name = request.Name,
            Email = request.Email,
            Password = HashPassword(request.Password)
        };

        var savedUser = _userRepository.Save(user);
        return MapToResponse(savedUser);
    }

    public UserResponse GetUser(string id)
    {
        var user = _userRepository.FindById(id);
        if (user == null)
        {
            throw new UserNotFoundException("User not found");
        }
        return MapToResponse(user);
    }

    public LoginResponse Login(LoginRequest request)
    {
        var user = _userRepository.FindByEmail(request.Email);
        if (user == null)
        {
            throw new AuthenticationException("Invalid email or password");
        }

        // Verify password
        var hashedPassword = HashPassword(request.Password);
        if (user.Password != hashedPassword)
        {
            throw new AuthenticationException("Invalid email or password");
        }

        // Generate simple token (in production, use JWT)
        var token = GenerateToken(user);

        return new LoginResponse
        {
            Token = token,
            User = MapToResponse(user)
        };
    }

    private UserResponse MapToResponse(User user)
    {
        return new UserResponse
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt
        };
    }

    private string HashPassword(string password)
    {
        using var sha256 = SHA256.Create();
        var hashedBytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(password));
        return Convert.ToBase64String(hashedBytes);
    }

    private string GenerateToken(User user)
    {
        // Simple token generation (in production, use JWT)
        var tokenData = $"{user.Id}:{user.Email}:{DateTime.UtcNow.Ticks}";
        var tokenBytes = Encoding.UTF8.GetBytes(tokenData);
        return Convert.ToBase64String(tokenBytes);
    }
}

File: api/services/monolith/src/main/java/com/example/monolith/user/service/UserService.java
package com.example.monolith.user.service;

import com.example.monolith.user.dto.CreateUserRequest;
import com.example.monolith.user.dto.LoginRequest;
import com.example.monolith.user.dto.LoginResponse;
import com.example.monolith.user.dto.UserResponse;
import com.example.monolith.user.exception.AuthenticationException;
import com.example.monolith.user.exception.UserAlreadyExistsException;
import com.example.monolith.user.exception.UserNotFoundException;
import com.example.monolith.user.model.User;
import com.example.monolith.user.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.UUID;

/**
 * Service layer for user management
 */
@Service
public class UserService {
    
    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }
    
    /**
     * Create a new user
     */
    public UserResponse createUser(CreateUserRequest request) {
        // Check if user already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new UserAlreadyExistsException("User with email " + request.getEmail() + " already exists");
        }
        
        // Create new user
        String userId = UUID.randomUUID().toString();
        User user = new User(userId, request.getName(), request.getEmail(), request.getPassword());
        User savedUser = userRepository.save(user);
        
        return mapToResponse(savedUser);
    }
    
    /**
     * Get user by ID
     */
    public UserResponse getUser(String id) {
        if (id == null || id.isBlank()) {
            throw new UserNotFoundException("User not found");
        }
        User user = userRepository.findById(id)
                .orElseThrow(() -> new UserNotFoundException("User not found"));
        
        return mapToResponse(user);
    }
    
    /**
     * Authenticate user
     */
    public LoginResponse authenticate(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new AuthenticationException("Invalid email or password"));
        
        if (!user.getPassword().equals(request.getPassword())) {
            throw new AuthenticationException("Invalid email or password");
        }
        
        // Generate simple token (in production, use proper JWT)
        String token = UUID.randomUUID().toString();
        String loginTime = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        
        return new LoginResponse(
            "Login successful",
            token,
            user.getId(),
            user.getName(),
            user.getEmail(),
            loginTime
        );
    }
    
    /**
     * Map User entity to UserResponse DTO
     */
    private UserResponse mapToResponse(User user) {
        String createdAtStr = user.getCreatedAt() != null 
            ? user.getCreatedAt().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME) 
            : null;
        return new UserResponse(
            user.getId(),
            user.getName(),
            user.getEmail(),
            createdAtStr
        );
    }
}
